### PR TITLE
Simplified Data Store interface

### PR DESCRIPTION
**Pull request by thehenrytsai**
1. Removed the need for SQL-like DB for reference counting.
2. Removed  the need for `associate()` implementation entirely.
3. Removed the need to return `dataCid` from data store and handled the `dataCid` computation within the `dwn-sdk-js` itself.
4. As a result, simplified the Level implementation which now only requires 135 lines of code.
5. Various other complexity that went away because of the simplification.

In summary an implementer now only needs to implement `put()`, `get()` and `delete()`.

### CODE REVIEW

**Review by LiranCohen — CHANGES_REQUESTED** [review]
Looks good!!! I love the simplification, makes reading a lot of things much easier, as well as some of the renames and comments you sprinkled in.

Just a couple small things.

File: src/store/data-store-level.ts [hunk 1]
```diff
@@ -1,25 +1,18 @@
...
  * A simple implementation of {@link DataStore} that works in both the browser and server-side.
  * Leverages LevelDB under the hood.
  *
- * It has the following structure (`+` represents a sublevel and `->` represents a key->value pair):
- *   'data' + <tenant> + <dataCid> -> <data>
- *   'references' + <tenant> + <dataCid> + <messageCid> -> PLACEHOLDER_VALUE
- *
- * This allows for the <data> to be shared for everything that uses the same <dataCid> while also making
- * sure that the <data> can only be deleted if there are no <messageCid> for any <tenant> still using it.
+ * It has the following structure (`+` represents an additional sublevel/partition):
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/data-store-level.ts:14]: I deliberately removed the part about `key->value pair`, because it is not correct. In `<dataCid> -> <data>`

The `dataCid` is still a partition. it's within that partition, there are a a lot of key-value pairs. Let me know if this doesn't make sense and how you'd like me to update the description given this context.

File: src/store/data-store-level.ts [hunk 2]
```diff
@@ -102,54 +84,21 @@ export class DataStoreLevel implements DataStore {
...
-    const blockstoreForReferenceCounting = await this.getBlockstoreForReferenceCounting(tenant, dataCid);
-    await blockstoreForReferenceCounting.delete(messageCid);
+    const partitionForRecordId = await this.getPartitionForRecordId(tenant, recordId);
 
-    const wasLastReference = await blockstoreForReferenceCounting.isEmpty();
-    if (!wasLastReference) {
-      return;
+    const noDataLeft = await partitionForRecordId.isEmpty();
+    if (noDataLeft) {
+      await blockstoreForData.clear();
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/data-store-level.ts]: Yes, thanks for the keen eyes! Though now I look at it again, I don't think the block has any point to it after the redesign (maybe even before?! can't be bother thinking about the old code), so removed it entirely. File is down to 120 lines, I'm sure we can do <100 lines if we really want to. 😆

File: src/types/data-store.ts [hunk 3]
```diff
@@ -15,39 +15,29 @@ export interface DataStore {
...
 
   /**
-   * Puts the given data in store.
-   * It is expected that the CID of the dataStream matches the given dataCid.
-   * The returned dataCid and returned dataSize will be verified against the given dataCid (and inferred dataSize).
-   * @param messageCid CID of the message that references the data.
-   * @returns The CID and size in number of bytes of the data stored.
+   * Stores the given data.
+   * @param recordId The logical ID of the record that references the data.
+   * @param dataCid The IPFS CID of the data.
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/types/data-store.ts:20]: I really meant it as IPFS CID as that's the hash we create using the IPFS importer. While an IPFS CID is an IPLD CID, I  used IPFS CID because:

1. It is arguably more generally known.
1. It is more specific.

File: src/types/data-store.ts [hunk 4]
```diff
@@ -56,23 +46,19 @@ export interface DataStore {
  * Result of a data store `put()` method call.
  */
 export type PutResult = {
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/types/data-store.ts:48]: Good call! Imma call it `DataStorePutResult` just to make the name longer. 🤣

File: src/types/data-store.ts [hunk 5]
```diff
@@ -56,23 +46,19 @@ export interface DataStore {
...
+   * The number of bytes of the data stored.
+   */
   dataSize: number;
 };
 
 /**
- * Result of a data store `get()` method call.
+ * Result of a data store `get()` method call if the data exists.
  */
 export type GetResult = {
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/types/data-store.ts:58]: Going for `DataStoreGetResult`.

File: src/store/data-store-level.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (LiranCohen) [reply src/store/data-store-level.ts:14]: Yeah that makes sense.

File: src/types/data-store.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LiranCohen) [reply src/types/data-store.ts:20]: Makes sense as it's more specific.

File: src/store/data-store-level.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LiranCohen) [reply src/store/data-store-level.ts]: Yeah I wasn't positive if it actually does anything since the partition is already empty.
